lib/Sema/CSDiagnostics.cpp

```diff
@@ -580,8 +580,21 @@ bool MissingOptionalUnwrapFailure::diagnoseAsError() {
     anchor = assignExpr->getSrc();
   
   auto *unwrapped = anchor->getValueProvidingExpr();
-  auto type = getType(anchor)->getRValueType();
```

Comment:
Yeah, that's true. Although, it breaks some tests. For example:

```swift
struct S {
  var x: Int = 0
  let y: Int = 0

  mutating func mutateS() {}

  init() {}
}

struct T {
  var mutS: S? = nil
  let immS: S? = nil

  mutating func mutateT() {}

  init() {}
}

var mutT: T?
_ = mutT?.mutS?.x + 0
```

It will give you a `binary operator '+' cannot be applied to operands of type 'Int?' and 'Int'` error.

```swift
class A { }
class B : A { 
  func iAmAB() {}
  func createB() -> B { return B() }
}

let bo: B? = B()
let b2: B = bo?.createB()
```

Here, it will give you a `cannot convert value of type 'B?' to specified type 'B'` error. 

This is why I added the extra checks.

Reply:
Right, and the reason for that is implicit conversion in cases like that. I think it would make sense move this logic to `diagnoseUnwrap` and if the sub-expression of the `OptionalEvaluationExpr` doesn't have `OptionalType` we should go ahead and make it optional instead of unwrapping.

Reply:
Actually I just realized that it might be better to detect that anchor is pointing to `OptionalEvaluationExpr` and just don't wrap in `diagnoseUnwrap`, that should be a better fix because that expression would always have a "correct" type.

Reply:
Here is what I mean:

```c++
  Type unwrappedType = isa<OptionalEvaluationExpr>(E) ? type : type->getOptionalObjectType();
  if (!unwrappedType)
    return false;
```

At the top of `diagnoseUnwrap` - because type of the `OptionalEvaluationExpr` is already the unwrapped one.

Reply:
And for that `type` you can just add one more level of optional on top.

Reply:
Another idea to try and simplify this might be to record both types in the fix and provide it to diagnostic, that might actually be the easiest way to go here.

Reply:
Oh! It stills not perfect though, as now it throws `value of optional type 'B?' must be unwrapped to a value of type 'B?'` for

```swift
class A { }
class B : A { 
  func iAmAB() {}
  func createB() -> B { return B() }
}

let bo: B? = B()
let b2: B = bo?.createB()
```

Seems like there are a few edge cases that trip the diagnostics.

Reply:
Okay, I will give that a try!

Reply:
@xedin Just wanted to clarify one thing - we you suggesting to store the base type and unwrapped type? For example, if the base type was `Foo??`, we would then store that and one level of unwrap (i.e. `Foo?`) and use that for the diagnostic?

I can find 5 instances of `ForceOptional::create` but in each case, I have access to a different set of types. For example, the one in `ConstraintSystem::matchTypes()` gives me a type variable wrapped in an optional in some cases (inside `if (forceUnwrapPossible) { ... }`) and I think that's the reason why the diagnostic would print `_?` or `_` as the type. I tried to use `getFixedType()` but that doesn't give me the right type.

I have pushed my changes if you want to take a look.

Reply:
Yes, I was thinking about saving original type with all of the optionals and the unwrapped result which is going to be used, you need to do `simplifyType` to access types type variables are bound to.

Reply:
@xedin Okay, that's exactly what I have done. It works for the above examples now. I ran tests and I have 5 failures. It seems like it has some troubles with some functions/patterns. Can you take a look at the code I pushed?

For example, it fails a test in `Constraints/patterns.swift`:

Expected error: `case .init(opt: 0): break // expected-error{{pattern cannot match values of type 'StaticMembers'}}`
Actual Error: `value of optional type 'StaticMembers?' must be unwrapped to a value of type 'StaticMembers'`

In that example, adding `?` to the case fixes the issue, so it can pattern match optional. Do you think this diagnostic is incorrect or does the test file need to be updated.

Reply:
Will take a look in a bit, thank you!